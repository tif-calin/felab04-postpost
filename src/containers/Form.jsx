import React, { useEffect } from 'react';
import { Buffer } from 'buffer';
import Payload from '../components/form/Payload';
import ParamsConfig from '../components/form/ParamsConfig';
import sendRequest from '../services/sendRequest.js';
import { addToHistory, getHistory } from '../services/localStorage.js';
import './Form.scss';

const fakeData = {
  url: 'https://www.onezoom.org/popularity/list',
  /*params: {
    key: 0,
    otts: 563159,
    expand_taxa: true,
    max: 12,
    names: true,
    include_raw: true,
    sort: 'standard'
  },*/
  params: {},
  request: {
    method: 'GET'
  }
};

export default function Form() {
  const [url, setUrl] = React.useState('');
  const [params, setParams] = React.useState({});
  const [request, setRequest] = React.useState({});
  const [auth, setAuth] = React.useState('');
  const [username, setUsername] = React.useState('');
  const [password, setPassword] = React.useState('');
  const [authBearerToken, setAuthBearerToken] = React.useState('');
  const [payload, setPayload] = React.useState();

  const [validJSON, setValidJSON] = React.useState(false);
  const [loading, setLoading] = React.useState(true);
  const [history, setHistory] = React.useState(getHistory());

  useEffect(() => {
    const previous = [...history].pop() || fakeData;

    setUrl(previous.url);
    setParams(previous.params);
    setRequest(previous.request);
    setLoading(false);
  }, []);

  useEffect(() => {
    if (!auth) {
      setUsername('');
      setPassword('');
    }
  }, [auth]);

  const handleBodyChange = e => {
    try { JSON.parse(e.target.value); }
    catch { 
      setValidJSON(false); 
      return;
    }

    setValidJSON(true);

    setRequest({ ...request, body: e.target.value });
  };

  const handleRequest = async e => {
    e.preventDefault();

    // set the body if relevant method
    if (['GET', 'DELETE'].includes(request.method)) {
      // eslint-disable-next-line no-unused-vars
      const { body: _, ...rest } = request;
      setRequest(rest);
    }

    // check for auth
    if (auth === 'basic') {
      const authHeader = 
        `${Buffer.from(username, 'base64')}:${Buffer.from(password, 'base64')}`
      ;
      const { headers } = request;
      headers['Authorization'] = `Basic ${authHeader}`;
      setRequest({ ...request, headers });
    } else if (auth === 'bearer') {
      const { headers } = request;
      headers['Authorization'] = `Bearer ${authBearerToken}`;
      setRequest({ ...request, headers });
    }

    // make the request, add to history if successful
    const resp = await sendRequest(url, params, request);
    if (resp.ok) {
      setPayload(await resp.json());
      addToHistory(url, params, request);
      setHistory(getHistory());
    }
  };

  return <div className="Form">
    <h2>postpost</h2>

    {!loading && <form onSubmit={handleRequest}>
      <fieldset name="url" onChange={e => setUrl(e.target.value)}>
        <legend>url</legend>
        <input 
          defaultValue={url}
          name="url"
          type="url" 
          placeholder="https://www.example.com/api/v1/fruit"
        />
        <button type="submit">send</button>
      </fieldset>

      <fieldset name="params">
        <legend>params</legend>
        <ParamsConfig params={params} setParams={setParams}/>
      </fieldset>

      <fieldset name="method" onChange={
        e => setRequest({ ...request, method: e.target.value })
      }>
        <legend>method</legend>
        <label>
          <input type="radio" value="GET" name="method" 
            defaultChecked={request.method === 'GET'}
          />
          <span>GET</span>
        </label>
        <label>
          <input type="radio" value="POST" name="method" 
            defaultChecked={request.method === 'POST'}
          />
          <span>POST</span>
        </label>
        <label>
          <input type="radio" value="PUT" name="method" 
            defaultChecked={request.method === 'PUT'}
          />
          <span>PUT</span>
        </label>
        <label>
          <input type="radio" value="DELETE" name="method" 
            defaultChecked={request.method === 'DELETE'}
          />
          <span>DELETE</span>
        </label>
        <label>
          <input type="radio" value="PATCH" name="method" 
            defaultChecked={request.method === 'PATCH'}
          />
          <span>PATCH</span>
        </label>
      </fieldset>

      <fieldset name="auth">
        <legend>auth</legend>
        <select 
          name="auth" 
          onChange={e => setAuth(e.target.value)}
          defaultValue={auth}
        >
          <option value="">none</option>
          <option value="basic">basic</option>
          <option hidden value="digest">digest</option>
          <option value="bearer">bearer</option>
        </select>
        {auth && <div>{auth === 'basic'
          ? <>
            <input 
              type="text" 
              name="username" 
              placeholder="username"
              onChange={e => setUsername(e.target.value)}
            />
            <input 
              type="text" 
              name="password" 
              placeholder="password"
              onChange={e => setPassword(e.target.value)}
            />
          </>
          : <input 
            type="text" 
            name="token" 
            placeholder="token"
            onChange={e => setAuthBearerToken(e.target.value)}
          />
        }</div>}
      </fieldset>

      <fieldset 
        onChange={handleBodyChange}
        style={{ 
          display: ['GET', 'DELETE'].includes(request.method) ? 'none' : '' 
        }}
      >
        <legend>request body</legend>
        <textarea 
          name="body" 
          placeholder="body" 
          defaultValue={JSON.stringify(
            request.body && JSON.parse(request.body), 
            null, 
            2
          )}
          style={{ backgroundColor: validJSON ? '#5f51' : '#d555' }}
        />
      </fieldset>
    </form>}

    <Payload payload={payload}/>

    {Boolean(history.length) && <><h3>history</h3>
      <div>
        {history.map(call => <p key={call.timestamp}>{call.url}</p>)}
      </div>
    </>}
  </div>;
}
