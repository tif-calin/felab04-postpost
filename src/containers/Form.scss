#root {
  max-width: 576px;
}

.Form {

input[type="radio"] {
  cursor: pointer;
  display: none;

  & + span {
    border: 1px solid #ccc0;
    border-radius: 5px;
    cursor: pointer;
    padding: 0.1rem 0.25rem;
  }

  &:checked + span {
    border: 1px solid #eee;
    background-color: #ddd;
  }
}

&> *:last-child {
  margin-top: 2rem;
}

form > fieldset {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;

  &[name="url"] {
    flex-direction: column;

    input {
      width: 100%;
    }

    > div {
      display: flex;
      justify-content: space-between;
    }
  }

  &[name="params"]{
    display: none;
  }

  &[name="method"] {
    flex-direction: row;
    justify-content: space-between;
  }

  &:last-of-type textarea {
    width: 100%;
    height: 5rem;
  }
}

table {
  width: 100%;
  text-align: center;

  tr {
    display: flex;
    justify-content: space-between;

    th {
      font-weight: normal;
    }

    td, th {
      width: 100%;
    }

    td button {
      width: 100%;
    }

    td:last-child span {
      display: flex;
    }
  }
}

}
