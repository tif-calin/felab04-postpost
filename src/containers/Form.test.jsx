import React from 'react';
import { render, cleanup } from '@testing-library/react';
import Form from './Form';

describe('Form component', () => {
  afterEach(() => cleanup());

  it('snpashot test Form', () => {
    const { asFragment } = render(<Form />);
    expect(asFragment()).toMatchSnapshot();
  });

});
