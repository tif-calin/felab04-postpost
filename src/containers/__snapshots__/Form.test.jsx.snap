// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component snpashot test Form 1`] = `
<DocumentFragment>
  <div
    class="Form"
  >
    <h2>
      postpost
    </h2>
    <form>
      <fieldset
        name="url"
      >
        <legend>
          url
        </legend>
        <input
          name="url"
          placeholder="https://www.example.com/api/v1/fruit"
          type="url"
          value="https://www.onezoom.org/popularity/list"
        />
        <button
          type="submit"
        >
          send
        </button>
      </fieldset>
      <fieldset
        name="params"
      >
        <legend>
          params
        </legend>
        <table
          class="ParamsConfig"
          colspan="3"
        >
          <thead>
            <tr>
              <th>
                key
              </th>
              <th>
                value
              </th>
              <th />
            </tr>
          </thead>
          <tbody>
            <tr
              class="ParamsEntry"
            >
              <td>
                <input
                  class="query-key"
                  disabled=""
                  placeholder="KEY"
                  type="text"
                  value="otts"
                />
              </td>
              <td>
                <input
                  class="query-val"
                  disabled=""
                  placeholder="VALUE"
                  type="text"
                  value="563159"
                />
              </td>
              <td>
                <span>
                  <button
                    type="button"
                  >
                    edit
                  </button>
                  <button
                    data-query-key="otts"
                    type="button"
                  >
                    remove
                  </button>
                </span>
              </td>
            </tr>
            <tr
              class="ParamsEntry"
            >
              <td>
                <input
                  class="query-key"
                  disabled=""
                  placeholder="KEY"
                  type="text"
                  value="expand_taxa"
                />
              </td>
              <td>
                <input
                  class="query-val"
                  disabled=""
                  placeholder="VALUE"
                  type="text"
                  value="true"
                />
              </td>
              <td>
                <span>
                  <button
                    type="button"
                  >
                    edit
                  </button>
                  <button
                    data-query-key="expand_taxa"
                    type="button"
                  >
                    remove
                  </button>
                </span>
              </td>
            </tr>
            <tr
              class="ParamsEntry"
            >
              <td>
                <input
                  class="query-key"
                  disabled=""
                  placeholder="KEY"
                  type="text"
                  value="max"
                />
              </td>
              <td>
                <input
                  class="query-val"
                  disabled=""
                  placeholder="VALUE"
                  type="text"
                  value="12"
                />
              </td>
              <td>
                <span>
                  <button
                    type="button"
                  >
                    edit
                  </button>
                  <button
                    data-query-key="max"
                    type="button"
                  >
                    remove
                  </button>
                </span>
              </td>
            </tr>
            <tr
              class="ParamsEntry"
            >
              <td>
                <input
                  class="query-key"
                  disabled=""
                  placeholder="KEY"
                  type="text"
                  value="names"
                />
              </td>
              <td>
                <input
                  class="query-val"
                  disabled=""
                  placeholder="VALUE"
                  type="text"
                  value="true"
                />
              </td>
              <td>
                <span>
                  <button
                    type="button"
                  >
                    edit
                  </button>
                  <button
                    data-query-key="names"
                    type="button"
                  >
                    remove
                  </button>
                </span>
              </td>
            </tr>
            <tr
              class="ParamsEntry"
            >
              <td>
                <input
                  class="query-key"
                  disabled=""
                  placeholder="KEY"
                  type="text"
                  value="include_raw"
                />
              </td>
              <td>
                <input
                  class="query-val"
                  disabled=""
                  placeholder="VALUE"
                  type="text"
                  value="true"
                />
              </td>
              <td>
                <span>
                  <button
                    type="button"
                  >
                    edit
                  </button>
                  <button
                    data-query-key="include_raw"
                    type="button"
                  >
                    remove
                  </button>
                </span>
              </td>
            </tr>
            <tr
              class="ParamsEntry"
            >
              <td>
                <input
                  class="query-key"
                  disabled=""
                  placeholder="KEY"
                  type="text"
                  value="sort"
                />
              </td>
              <td>
                <input
                  class="query-val"
                  disabled=""
                  placeholder="VALUE"
                  type="text"
                  value="standard"
                />
              </td>
              <td>
                <span>
                  <button
                    type="button"
                  >
                    edit
                  </button>
                  <button
                    data-query-key="sort"
                    type="button"
                  >
                    remove
                  </button>
                </span>
              </td>
            </tr>
            <tr
              class="ParamsEntry"
            >
              <td>
                <input
                  class="query-key"
                  placeholder="KEY"
                  type="text"
                  value=""
                />
              </td>
              <td>
                <input
                  class="query-val"
                  placeholder="VALUE"
                  type="text"
                  value=""
                />
              </td>
              <td>
                <button
                  type="button"
                >
                  save
                </button>
              </td>
            </tr>
          </tbody>
        </table>
      </fieldset>
      <fieldset
        name="method"
      >
        <legend>
          method
        </legend>
        <label>
          <input
            checked=""
            name="method"
            type="radio"
            value="GET"
          />
          <span>
            GET
          </span>
        </label>
        <label>
          <input
            name="method"
            type="radio"
            value="POST"
          />
          <span>
            POST
          </span>
        </label>
        <label>
          <input
            name="method"
            type="radio"
            value="PUT"
          />
          <span>
            PUT
          </span>
        </label>
        <label>
          <input
            name="method"
            type="radio"
            value="DELETE"
          />
          <span>
            DELETE
          </span>
        </label>
        <label>
          <input
            name="method"
            type="radio"
            value="PATCH"
          />
          <span>
            PATCH
          </span>
        </label>
      </fieldset>
      <fieldset
        name="auth"
      >
        <legend>
          auth
        </legend>
        <select
          name="auth"
        >
          <option
            selected=""
            value=""
          >
            none
          </option>
          <option
            value="basic"
          >
            basic
          </option>
          <option
            hidden=""
            value="digest"
          >
            digest
          </option>
          <option
            value="bearer"
          >
            bearer
          </option>
        </select>
      </fieldset>
      <fieldset
        style="display: none;"
      >
        <legend>
          request body
        </legend>
        <textarea
          name="body"
          placeholder="body"
          style="background-color: rgba(221, 85, 85, 0.333);"
        />
      </fieldset>
    </form>
    <div
      class="Payload"
    >
      <fieldset>
        <legend>
          response body
        </legend>
        <pre>
          <code />
        </pre>
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;
